bundles/org.openhab.binding.upb/src/main/java/org/openhab/binding/upb/handler/UPBThingHandler.java

```diff
@@ -0,0 +1,292 @@
+            logger.warn("invalid network ID {} for {}", val, getThing().getUID());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "invalid network ID");
+            return;
+        } else {
+            networkId = val.byteValue();
+        }
+
+        final BigDecimal cfgUnitId = (BigDecimal) getConfig().get(Constants.CONFIGURATION_UNIT_ID);
+        if (cfgUnitId == null) {
+            logger.warn("Unit ID is not set in {}", getThing().getUID());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "missing unit ID");
+            return;
+        }
+        unitId = cfgUnitId.intValue();
+        if (unitId < 1 || unitId > 250) {
+            logger.warn("Unit ID ({}) out of range for {}", cfgUnitId, getThing().getUID());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "invalid unit ID");
+            return;
+        }
+
+        final Bridge bridge = getBridge();
+        if (bridge == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE, Constants.OFFLINE_CTLR_OFFLINE);
+            return;
+        }
+        bridgeStatusChanged(bridge.getStatusInfo());
+    }
+
+    @Override
+    public void bridgeStatusChanged(final ThingStatusInfo bridgeStatusInfo) {
+        logger.debug("DEV {}: Controller status is {}", unitId, bridgeStatusInfo.getStatus());
+
+        if (bridgeStatusInfo.getStatus() != ThingStatus.ONLINE) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE, Constants.OFFLINE_CTLR_OFFLINE);
+            return;
+        }
+
+        logger.debug("DEV {}: Controller is ONLINE. Starting device initialisation.", unitId);
+
+        final Bridge bridge = getBridge();
+        if (bridge == null) {
+            logger.debug("DEV {}: bridge is null!", unitId);
+            return;
+        }
+        final PIMHandler bridgeHandler = (PIMHandler) bridge.getHandler();
+        if (bridgeHandler == null) {
+            logger.debug("DEV {}: bridge handler is null!", unitId);
+            return;
+        }
+        updateDeviceStatus(bridgeHandler);
+        pingDevice();
+    }
+
+    @Override
+    public void handleCommand(final ChannelUID channelUID, final Command cmd) {
+        final PIMHandler pimHandler = getPIMHandler();
+        if (pimHandler == null) {
+            logger.info("DEV {}: received cmd {} but no bridge handler", unitId, cmd);
+            return;
+        }
+
+        final MessageBuilder message;
+        if (cmd == OnOffType.ON) {
+            message = MessageBuilder.forCommand(ACTIVATE);
+        } else if (cmd == OnOffType.OFF) {
+            message = MessageBuilder.forCommand(DEACTIVATE);
+        } else if (cmd instanceof PercentType) {
+            message = MessageBuilder.forCommand(GOTO).args(((PercentType) cmd).byteValue());
+        } else if (cmd == RefreshType.REFRESH) {
+            refreshDeviceState();
```

Comment:
The refresh is given for each channel. So calling refreshDeviceState might be a bit of an overload.

Reply:
`refreshDeviceState` might still cause multiple calls to the device on refresh. You could use a (org.eclipse..) `ExpiringCache` with something of 3 seconds timeout to avoid it when the ui calls refresh on all channels.